### PR TITLE
Shard-level const column

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now functions can be shard-level constants, which means if it's executed in the context of some distributed table, it generates a normal column, otherwise it produces a constant value. Notable functions are: `hostName()`, `tcpPort()`, `version()`, `buildId()`, `uptime()`, etc.

This PR also introduces two functions: `shardNum()` and `shardCount()`. They are also shard-level constants and provide shard info during query execution.

The old `_shard_num` virtual column will still work but should be deprecated.

Detailed description / Documentation draft:
.

### CODE REVIEW

File: src/Functions/buildId.cpp [hunk 1]
```diff
@@ -42,6 +47,9 @@ class FunctionBuildId : public IFunction
     {
         return DataTypeString().createColumnConst(input_rows_count, SymbolIndex::instance()->getBuildIDHex());
     }
+
+private:
+    ContextPtr context;
```
Comment (KochetovNicolai) [inline src/Functions/buildId.cpp:52]: Let's better store only `is_distributed` flag instead of whole context.

**Review by KochetovNicolai — APPROVED** [review]
Awesome!

File: src/Functions/buildId.cpp (hunk 1 above)
Comment (amosbird) [reply src/Functions/buildId.cpp:52]: Done

File: src/Functions/getScalar.cpp [hunk 2]
```diff
@@ -60,11 +61,89 @@ class FunctionGetScalar : public IFunction, WithContext
     mutable ColumnWithTypeAndName scalar;
 };
 
+
+/** Get special scalar values
+  */
+template <typename Scalar>
+class FunctionGetSpecialScalar : public IFunction, WithContext
```
Comment (KochetovNicolai) [inline src/Functions/getScalar.cpp:68]: Looks like we don't actually need `WithContext` here.
Context is used only to get scalar in the constructor

Comment (amosbird) [reply src/Functions/getScalar.cpp:68]: Indeed.

**Review by KochetovNicolai — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
Fast test is ok.
Not waiting for other checks, cause prev commit is green and the change is tiny

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-14057.

**Pull request merged**